torch/csrc/jit/tensorexpr/ir_serializer.cpp

```diff
@@ -0,0 +1,557 @@
+
+void IRSerializer::visit(const Sub* v) {
+  visitBinaryOp(v, "Sub", this);
+}
+
+void IRSerializer::visit(const Mul* v) {
+  visitBinaryOp(v, "Mul", this);
+}
+
+void IRSerializer::visit(const Div* v) {
+  visitBinaryOp(v, "Div", this);
+}
+
+void IRSerializer::visit(const Mod* v) {
+  visitBinaryOp(v, "Mod", this);
+}
+
+void IRSerializer::visit(const Max* v) {
+  visitBinaryOp(v, "Max", this);
+}
+
+void IRSerializer::visit(const Min* v) {
+  visitBinaryOp(v, "Min", this);
+}
+
+void IRSerializer::visit(const And* v) {
+  visitBinaryOp(v, "And", this);
+}
+
+void IRSerializer::visit(const Or* v) {
+  visitBinaryOp(v, "Or", this);
+}
+
+void IRSerializer::visit(const Xor* v) {
+  visitBinaryOp(v, "Xor", this);
+}
+
+void IRSerializer::visit(const Lshift* v) {
+  visitBinaryOp(v, "Lshift", this);
+}
+
+void IRSerializer::visit(const Rshift* v) {
+  visitBinaryOp(v, "Rshift", this);
+}
+
+void IRSerializer::visit(const CompareSelect* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "CompareSelect"},
+          {"dtype", serializeDtype(v->dtype())},
+          {"lhs", visitAndSerialize(v->lhs())},
+          {"rhs", visitAndSerialize(v->rhs())},
+          {"ret_val1", visitAndSerialize(v->ret_val1())},
+          {"ret_val2", visitAndSerialize(v->ret_val2())},
+          {"CompareSelectOperation", v->compare_select_op()},
+      });
+}
+
+void IRSerializer::visit(const Cast* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {{"expr_type", "Cast"},
+       {"dtype", serializeDtype(v->dtype())},
+       {"src_value", visitAndSerialize(v->src_value())}});
+}
+
+void IRSerializer::visit(const Var* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {{"expr_type", "Var"},
+       {"dtype", serializeDtype(v->dtype())},
+       {"name", name_manager_.get_unique_name(v)}});
+}
+
+void IRSerializer::visit(const Ramp* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"expr_type", "Ramp"},
+          {"dtype", serializeDtype(v->dtype())},
+          {"base", visitAndSerialize(v->base())},
+          {"stride", visitAndSerialize(v->stride())},
+          {"lanes", v->lanes()},
+      });
+}
+
+void IRSerializer::visit(const Load* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"base_handle", visitAndSerialize(v->base_handle())},
+          {"indices", visitAndSerialize(v->indices())},
+          {"mask", visitAndSerialize(v->mask())},
+          {"stmt_type", "Load"},
+      });
+}
+
+void IRSerializer::visit(const Store* v) {
+  JSON_CACHE_GUARD();
+  putJSON(
+      v,
+      {
+          {"stmt_type", "Store"},
+          {"indices", visitAndSerialize(v->indices())},
+          {"value", visitAndSerialize(v->value())},
+          {"mask", visitAndSerialize(v->mask())},
+          {"buf", visitAndSerialize(v->buf())},
+      });
+}
+
+void IRSerializer::visit(const Block* v) {
+  JSON_CACHE_GUARD();
+  // TODO: why is Block using std::list instead of std::vector ?
```

Comment:
Because we often prepend Stmts to Blocks, or splice them, etc. I think that's why, anyway.

That said, it should probably be a std::deque.

Reply:
Yeah, correct. `std::list` is better than `std::vector` (and also than `std::deque`) when we have operations like `insertBefore` and `insertAfter`.

Reply:
As an aside @ZolotukhinM, why is list better than deque in that instance? The overhead of changing both sides of a doubly linked list doesn't seem significant at the N we're likely to see here. Or is there something about the implementation of deque I'm not aware of?

Reply:
The difference is not about inserting an item to the beginning or the end of the list - it's about inserting into the middle. I thought `deque` didn't allow to do that easly? Am I mistaken?

Reply:
@ZolotukhinM I found this looking around https://www.codeproject.com/Articles/5425/An-In-Depth-Study-of-the-STL-Deque-Container which was helpful in understanding deque.

From my reading, `deque` is a nice wrapper around `list<vector>` - gives you a midpoint implementation which supports random access but doesn't punish insertion as badly. I'd be curious about which is better for our purposes here - we don't actually allow random access of the stmt list, only iteration. Would it be helpful to find the nth stmt?

Reply:
> From my reading, deque is a nice wrapper around list<vector>
What do you mean by `list<vector>`? Did you mean `list<Stmt>`?

If I understood correctly, the article compares `deque` with `vector` - when compare these two the difference can be illustrated by a simple insertion (anywhere, except at the end). I.e. if we insert in the beginning of a block `deque` would be much better than `vector`, since complexity of the operation would be O(1) rather than O(N).

When we compare `deque` with `list`, the difference comes into play when we try to insert in the middle of the block. According to https://en.cppreference.com/w/cpp/container/deque/insert, complexity of insertion to the middle would be "Constant plus linear in the lesser of the distances between pos and either of the ends of the container." Do we really need to ever insert in the middle? Yes, we do - imagine we have a block with N loops and we split each of them with tail. During each of the splits, we need to insert the tail loop right after the given loop:
https://github.com/pytorch/pytorch/blob/822efb7275939902fab53ef578c0e6ef5f63e580/torch/csrc/jit/tensorexpr/loopnest.cpp#L1031
If we use `deque` (or `vector` fwiw), the complexity of this pass would be worse than if we use `list`.

Do I somehow read it wrong? Also, what is a benefit of using `deque` vs `list`?

Reply:
Otoh insertion into a vector of trivially movable types (like Stmt*) can be done with memmove, whereas insertion into a `list` will allocate (always, I think)... but now I'm just being a pain 😈